libs/entitlement/asm-provider/src/main/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterImpl.java

```diff
@@ -82,13 +91,63 @@ static ClassFileInfo getClassFileInfo(Class<?> clazz) throws IOException {
         return new ClassFileInfo(fileName, originalBytecodes);
     }
 
+    private enum VerificationPhase {
+        BEFORE_INSTRUMENTATION,
+        AFTER_INSTRUMENTATION
+    }
+
+    private static String verify(byte[] classfileBuffer) {
+        ClassReader reader = new ClassReader(classfileBuffer);
+        StringWriter stringWriter = new StringWriter();
+        PrintWriter printWriter = new PrintWriter(stringWriter);
+        CheckClassAdapter.verify(reader, false, printWriter);
+        return stringWriter.toString();
+    }
+
+    private static void verifyAndLog(byte[] classfileBuffer, String className, VerificationPhase phase) {
+        var failureLogLevel = (phase == VerificationPhase.BEFORE_INSTRUMENTATION ? Level.WARN : Level.ERROR);
```

Comment:
Why the different log levels?

Reply:
I assumed we want to highlight if something fails verification after our instrumentation, it is something we want to pay immediate attention. But it's also true that something that fails verification before will inevitably fail after too.
I thought about making this compare the error before/after, and emit a WARN if they are identical (not our fault), and ERROR if different (our fault)

Reply:
If the incoming bytecode fails verification, we should just bail early with an AssertionError, we shouldn't continue on, so the log levels still don't make sense to me. In either case, it should be an error (not just logged).

Reply:
> Some JDK classes fail verification as they are

It's all linked to this: we can log, but we cannot fail.